File: exercice4_31.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/wait.h>
#include <fcntl.h>
#include <unistd.h>


int main(int argc, char const *argv[])
{
	int fd ;
	pid_t fils ;
	int tube[2] ;

	pipe(tube) ;
	fils = fork() ;

	if (fils == 0)
	{
		char nomfich[50] ;
		close(tube[1]) ;
		read(tube[0], nomfich, sizeof(nomfich)) ;
		close(tube[0]) ;

		char *arg[] = {"exercice4_32", nomfich, NULL} ;
		execvp("./exercice4_32", arg) ;
	}
	else if (fils > 0)
	{
		char *nomfich = "/tmp/test.txt" ;
		close(tube[0]) ;
		write(tube[1], nomfich, sizeof(nomfich)) ;
		close(tube[1]) ;
		

		FILE *fp ;
		
		if (mkfifo(nomfich, 0644) != 0)
		{
			perror("Problème de création du noeud de tube") ;
			exit (1) ;
		}

		fd = open(nomfich, O_WRONLY) ;
		fp = fdopen(fd, "w") ;
		fprintf(fp, "coucou\n") ;

		wait(NULL) ;

		//unlink(nomfich) ;

	}
	
	return 0;
}

File: Python/exercice4_31.py
import os
import sys
import time

def main():
    tube = os.pipe()
    
    pid = os.fork()
    
    if pid == 0:
        os.close(tube[1])  # Fermer l'extrémité écriture
        
        nomfich_bytes = os.read(tube[0], 50)
        nomfich = nomfich_bytes.decode().rstrip('\x00')
        os.close(tube[0])
        os.execvp("python3", ["python3", "exercice4_32.py", nomfich])
        
    else:
        nomfich = "/tmp/test.txt"
        
        os.close(tube[0])  # Fermer l'extrémité lecture
        os.write(tube[1], nomfich.encode())
        os.close(tube[1])
        
        try:
            os.mkfifo(nomfich, 0o644)
        except FileExistsError:
            print(f"Le FIFO {nomfich} existe déjà")
        except OSError as e:
            print(f"Erreur création FIFO: {e}")
            sys.exit(1)
        
        fd = os.open(nomfich, os.O_WRONLY)
        
        message = "coucou\n"
        os.write(fd, message.encode())
        os.close(fd)
        
        os.wait()
        
        #os.unlink(nomfich)

if __name__ == "__main__":
    main()